test/init/initialize.test.js

```diff
@@ -0,0 +1,101 @@
+/* eslint-disable jest/prefer-strict-equal */
+const { promises: fs } = require('fs');
+const initUtils = require('../../dist/src/cmds/init/initutils');
+const readlineUtils = require('../../dist/src/utils/readline');
+const miscUtils = require('../../dist/src/utils/misc');
+const { initHandler } = require('../../dist/src/cmds/init/initialize');
+
+describe('initialize', () => {
+  describe('initHandler', () => {
+
+    afterEach(() => {
+      jest.clearAllMocks();
+      jest.restoreAllMocks();
+    });
+
+    const mockPackage = {
+      name: 'foo',
+      version: '1.0.0',
+      description: '',
+      main: 'index.js',
+      scripts: { test: 'echo "Error: no test specified" && exit 1' },
+      author: '',
+      license: 'ISC',
+    };
+
+    const mockArgv = { foo: 'bar' };
+    const mockWallet = [
+      {
+        bundle: {},
+        initialPermissions: {},
+      },
+      { dist: 'dist', outfileName: 'bundle.js', port: 8081 },
+    ];
+    const expectedReturnValue = {
+      foo: 'bar',
+      src: 'index.js',
+      dist: 'dist',
+      outfileName: 'bundle.js',
+      port: 8081,
+    };
+
+    it('function successfully executes under normal conditions', async () => {
+      jest.spyOn(console, 'log').mockImplementation();
+      const fsWriteMock = jest.spyOn(fs, 'writeFile').mockImplementation(() => true);
+      const asyncPackageInitMock = jest.spyOn(initUtils, 'asyncPackageInit').mockImplementation(() => mockPackage);
+      const validateEmptyDirMock = jest.spyOn(initUtils, 'validateEmptyDir').mockImplementation();
+      const buildWeb3WalletMock = jest.spyOn(initUtils, 'buildWeb3Wallet').mockImplementation(() => mockWallet);
+      const closePromptMock = jest.spyOn(readlineUtils, 'closePrompt').mockImplementation(() => mockPackage);
+      expect(await initHandler(mockArgv)).toStrictEqual(expectedReturnValue);
+      expect(global.console.log).toHaveBeenCalledTimes(6);
+      expect(fsWriteMock).toHaveBeenCalledTimes(4);
+      expect(asyncPackageInitMock).toHaveBeenCalledTimes(1);
+      expect(validateEmptyDirMock).toHaveBeenCalledTimes(1);
+      expect(closePromptMock).toHaveBeenCalledTimes(1);
+      expect(buildWeb3WalletMock).toHaveBeenCalledWith(mockArgv);
```

Comment:
This is a general point applicable to all test cases in this file, although this case illustrates it well.

Recall that we are interested in testing _behavior_ with as little regard for implementation details as possible. To test the behavior of a function, we check whether its outputs matches our expectations.

`initHandler` has two outputs, (1) its return value, which we are already checking, and (2) what it writes to disk, which we are not checking. Meanwhile, we are checking how many times various internal functions are called, which is not necessary unless we want to check that `process.exit` was called to confirm that the function would have returned early.

Rather than checking how many times any of the intermediate mocks have been called, we should check what `fsWriteMock` was called with every time it was called, and see that it matches our expectations.

Reply:
lolz https://github.com/MetaMask/snaps-cli/commit/c95f97f7ad3def6ee74dfbf8f1269aff51288867#diff-6002baa6883b76d5318563e7608356cf38302408511e5a142502bb621fa94e08

Reply:
esp line 15 👀

Reply:
Also @rekmarks just wanted to check that this function behavior is as intended? 

```
  // write main js entry file
  const { main } = pkg;
  if (main !== undefined) {
    newArgs.src = main;
    try {
      await fs.writeFile(main, template.js);
      console.log(`Init: Wrote main entry file '${main}'`);
    } catch (err) {
      logError(`Init Error: Fatal: Failed to write main .js file '${main}'`, err);
      process.exit(1);
    }
  }
```
If not given a main field, the return value will not have a src field.